Library/Homebrew/dependency.rb

```diff
@@ -124,11 +124,39 @@ def merge_repeats(all)
       all.map(&:name).uniq.map do |name|
         deps = grouped.fetch(name)
         dep  = deps.first
-        tags = deps.flat_map(&:tags).uniq
+        tags = merge_tags(deps)
         option_names = deps.flat_map(&:option_names).uniq
         dep.class.new(name, tags, dep.env_proc, option_names)
       end
     end
+
+    private
+
+    def merge_tags(deps)
+      options = deps.flat_map(&:option_tags).uniq
+      merge_necessity(deps) + merge_temporality(deps) + options
+    end
+
+    def merge_necessity(deps)
+      if deps.any? { |dep| !dep.recommended? && !dep.optional? }
```

Comment:
It's kinda related to the `FIXME` comment below. Ideally, I would be just using `deps.any?(&:required?)` instead of the block here, but because of the (debatable) definition of `required?` I'm doing it like this (and trying to keep it as close as possible to the definition of `required?`).